File: .claude/hooks/rust-check.py
#!/usr/bin/env python3
"""
PostToolUse hook — runs cargo fmt + cargo clippy after any Edit/Write to a .rs file.
Exits with 2 if clippy reports errors, which triggers an asyncRewake notification.
"""
import json
import os
import subprocess
import sys

data = json.load(sys.stdin)
file_path = (data.get("tool_input") or {}).get("file_path", "")

# Only process Rust source files
if not file_path.endswith(".rs"):
    sys.exit(0)

# Normalize path (handles Windows backslashes)
file_path = os.path.normpath(file_path)

# Walk up the directory tree to find the nearest Cargo.toml
search_dir = os.path.dirname(os.path.abspath(file_path))
cargo_root = None
while True:
    if os.path.exists(os.path.join(search_dir, "Cargo.toml")):
        cargo_root = search_dir
        break
    parent = os.path.dirname(search_dir)
    if parent == search_dir:  # reached filesystem root
        break
    search_dir = parent

if cargo_root is None:
    sys.exit(0)

# 1. Format — always silent-succeeds
subprocess.run(["cargo", "fmt", "--all"], cwd=cargo_root)

# 2. Clippy — capture output so it can be forwarded on failure
r = subprocess.run(
    ["cargo", "clippy", "--all-targets"],
    cwd=cargo_root,
    capture_output=True,
    text=True,
)
# Print to terminal regardless
sys.stdout.write(r.stdout)
sys.stderr.write(r.stderr)
sys.stdout.flush()
sys.stderr.flush()

# Exit 2 triggers asyncRewake — Claude sees clippy output as a system-reminder
if r.returncode != 0:
    sys.exit(2)

File: mcp/crates/mcp-hooks/src/rust_check.rs
//! Post-edit Rust quality hook: `cargo fmt` + `cargo clippy`.

use anyhow::Result;
use serde_json::Value;
use std::io::{self, Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

/// Run `cargo fmt --all` and `cargo clippy --all-targets` for the Cargo
/// workspace that contains the file described in the hook JSON payload on stdin.
///
/// Reads a Claude Code `PostToolUse` JSON event from stdin. Does nothing if the
/// modified file is not a `.rs` file or if no `Cargo.toml` ancestor is found.
/// Exits with code 2 if clippy reports errors.
///
/// # Errors
///
/// Returns an error if stdin cannot be read or if the clippy subprocess cannot
/// be spawned.
pub fn run() -> Result<()> {
    let mut stdin = String::new();
    io::stdin().read_to_string(&mut stdin)?;

    let data: Value = serde_json::from_str(&stdin).unwrap_or(Value::Null);
    let file_path = data
        .get("tool_input")
        .and_then(|v| v.get("file_path"))
        .and_then(|v| v.as_str())
        .unwrap_or("");

    if !file_path.ends_with(".rs") {
        return Ok(());
    }

    let abs = PathBuf::from(file_path);
    let Some(cargo_root) = find_cargo_root(&abs) else {
        return Ok(());
    };

    // 1. cargo fmt — silent, always succeeds
    let _ = Command::new("cargo")
        .args(["fmt", "--all"])
        .current_dir(&cargo_root)
        .status();

    // 2. cargo clippy — capture output, forward on failure
    let out = Command::new("cargo")
        .args(["clippy", "--all-targets"])
        .current_dir(&cargo_root)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()?;

    io::stdout().write_all(&out.stdout)?;
    io::stderr().write_all(&out.stderr)?;
    io::stdout().flush()?;
    io::stderr().flush()?;

    if !out.status.success() {
        std::process::exit(2);
    }
    Ok(())
}

fn find_cargo_root(file: &Path) -> Option<PathBuf> {
    let mut dir = file.parent()?.to_path_buf();
    if !dir.is_absolute() {
        dir = std::env::current_dir().ok()?.join(&dir);
    }
    loop {
        if dir.join("Cargo.toml").exists() {
            return Some(dir);
        }
        if !dir.pop() {
            return None;
        }
    }
}
